mono/metadata/marshal.c

```diff
@@ -3264,6 +3265,88 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
 	}
 }
 
+static void
+mono_marshal_set_signature_callconv_from_attribute (MonoMethodSignature* sig, MonoType* cmod_type, MonoError* error)
+{
+	g_assert(cmod_type->type == MONO_TYPE_CLASS);
+	MonoClass* cmod_klass = mono_class_from_mono_type_internal (cmod_type);
+	g_assert (m_class_get_image (cmod_klass) == mono_defaults.corlib);
+	g_assert (!strcmp (m_class_get_name_space (cmod_klass), "System.Runtime.CompilerServices"));
+
+	const char* name = m_class_get_name (cmod_klass);
+	g_assert (g_str_has_prefix(name, "CallConv"));
+
+	name += strlen ("CallConv");
+	if (!strcmp (name, "Cdecl"))
+		sig->call_convention = MONO_CALL_C;
+	else if (!strcmp (name, "Stdcall"))
+		sig->call_convention = MONO_CALL_STDCALL;
+	else if (!strcmp (name, "Thiscall"))
+		sig->call_convention = MONO_CALL_THISCALL;
+	else if (!strcmp (name, "Fastcall"))
+		sig->call_convention = MONO_CALL_FASTCALL;
+	else if (!strcmp (name, "SuppressGCTransition"))
+		sig->suppress_gc_transition = 1;
+	// TODO: Support CallConvMemberFunction?
+	// TODO: Support multiple calling convetions?
+	//		 - Switch MonoCallConvention enum values to powers of 2
+	//		 - Adjust the code above with 'anding' the attribute parameter value
+}
+
+static void
+mono_marshal_set_callconv_from_unmanaged_callers_only_attribute (MonoMethod* method, MonoMethodSignature* csig)
+{
+#ifdef ENABLE_NETCORE
+	MonoClass* attr_class = mono_class_try_get_unmanaged_callers_only_attribute_class ();
+	if (!attr_class)
+		return;
+#endif
+
+	ERROR_DECL(error);
+	MonoCustomAttrInfo* cinfo = mono_custom_attrs_from_method_checked (method, error);
+	if (!is_ok(error) || !cinfo) {
+		mono_error_cleanup(error);
```

Comment:
Doesn't look like `mono_error_cleanup` is called if we get past here.

Reply:
I think that only needs called when an error happens. I have added checks after each function taking an error, and just `mono_error_assert_ok` for now. I'd rather crash and detect we are doing something wrong I think.